sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/UnevaluableAggs.scala

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions.aggregate
+
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.types._
+
+abstract class UnevaluableBooleanAggBase(arg: Expression)
```

Comment:
why don't we reuse `RuntimeReplaceable`?

Reply:
@mgaido91 RuntimeReplaceble works for scalar expressions but not for aggregate expressions.

Reply:
why it doesn't work? Sorry if the question is dumb, but I can't see the problem in using it here, probably I am missing something.

Reply:
@mgaido91 Actually i tried a few different ideas. The are in the comment of original PR [link](https://github.com/apache/spark/pull/22047).  I had prepared two branches with two approaches .. [comment] (https://github.com/apache/spark/pull/22047#discussion_r225700828). Could you please take a look ?

Reply:
@mgaido91 I tried it [here](https://github.com/apache/spark/compare/master...dilipbiswal:SPARK-19851-rewrite?expand=1). I had trouble getting it to work for window expressions. Thats why @cloud-fan suggested to try the current approach in this [comment](https://github.com/apache/spark/pull/22797#issuecomment-432106729)

Reply:
@mgaido91 I re-read your comments again and here is what i think you are implying : 
   1. In UnevaluableAggregate, define a method say `replacedExpression`
   2. The sub-class overrides it and return an actual rewritten expression
   3. In optimized, match UnevaluableAggregate and call `replacedExpression`

Did i understand it correctly ? If so, actually, i wouldn't prefer that way. The reason is, the replacedExpression is hidden from analyzer and so it may not be safe. The way ReplaceExpression framework is nicely designed, the analyzer resolves the rewritten expression normally (as its the child expression). Thats the reason, i have opted for specific/targeted rewrites. 

If Wenchen and you think otherwise, then i can change.

Reply:
yes, but in the analyzer you analyze the children of the current expression right? So we would just have something like `def replacedExpression = Min(arg)`, which means doing exactly the same which is done now, the only difference is where the conversion logic is put. Adn IMHO having all the conversion logic for all the expression in `ReplaceExpressions` is harder to maintain than having all the logic related to the expression contained in it.

Anyway, since you have a different opinion, let's see what @cloud-fan thinks on this. Thanks.

Reply:
We can leave a TODO saying that we should create a framework to replace aggregate functions, but I think the current patch is good enough for these 3 functions, and I'm not aware of more functions like them that we need to deal with.

Reply:
@cloud-fan @mgaido91 Thank you. I have added a TODO for now.